pkg/block/blocktest/adapter.go

```diff
@@ -0,0 +1,281 @@
+	}
+	dst := block.ObjectPointer{
+		StorageNamespace: storageNamespace,
+		Identifier:       "export/to/dst",
+		IdentifierType:   block.IdentifierTypeRelative,
+	}
+
+	require.NoError(t, adapter.Put(ctx, src, 0, strings.NewReader(contents), block.PutOpts{}))
+
+	require.NoError(t, adapter.Copy(ctx, src, dst))
+	reader, err := adapter.Get(ctx, dst, 0)
+	require.NoError(t, err)
+	got, err := io.ReadAll(reader)
+	require.NoError(t, err)
+	require.Equal(t, contents, string(got))
+}
+
+func testAdapterRemove(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	const content = "Content used for testing"
+	tests := []struct {
+		name              string
+		additionalObjects []string
+		path              string
+		wantErr           bool
+		wantTree          []string
+	}{
+		{
+			name:     "test_single",
+			path:     "README",
+			wantErr:  false,
+			wantTree: []string{},
+		},
+
+		{
+			name:     "test_under_folder",
+			path:     "src/tools.go",
+			wantErr:  false,
+			wantTree: []string{},
+		},
+		{
+			name:     "test_under_multiple_folders",
+			path:     "a/b/c/d.txt",
+			wantErr:  false,
+			wantTree: []string{},
+		},
+		{
+			name:              "file_in_the_way",
+			path:              "a/b/c/d.txt",
+			additionalObjects: []string{"a/b/blocker.txt"},
+			wantErr:           false,
+			wantTree:          []string{"/a/b/blocker.txt"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// setup env
+			envObjects := append(tt.additionalObjects, tt.path)
+			for _, p := range envObjects {
+				obj := block.ObjectPointer{
+					StorageNamespace: storageNamespace,
+					Identifier:       tt.name + "/" + p,
+					IdentifierType:   block.IdentifierTypeRelative,
+				}
+				require.NoError(t, adapter.Put(ctx, obj, 0, strings.NewReader(content), block.PutOpts{}))
+			}
+
+			// test Remove
+			obj := block.ObjectPointer{
+				StorageNamespace: storageNamespace,
+				Identifier:       tt.name + "/" + tt.path,
+				IdentifierType:   block.IdentifierTypeRelative,
+			}
+			if err := adapter.Remove(ctx, obj); (err != nil) != tt.wantErr {
+				t.Errorf("Remove() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			tree := dumpPathTree(t, ctx, adapter, storageNamespace, tt.name)
+			if diff := deep.Equal(tt.wantTree, tree); diff != nil {
+				t.Errorf("Remove() tree diff = %s", diff)
+			}
+		})
+	}
+}
+
+func dumpPathTree(t testing.TB, ctx context.Context, adapter block.Adapter, storageNamespace, root string) []string {
+	t.Helper()
+	tree := make([]string, 0)
+
+	err := adapter.Walk(ctx, block.WalkOpts{
+		StorageNamespace: storageNamespace,
+		Prefix:           root,
+	}, func(id string) error {
+		_, p, _ := strings.Cut(id, root)
+		tree = append(tree, p)
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("walking on '%s': %s", root, err)
+	}
+	sort.Strings(tree)
+	return tree
+}
+
+func testAdapterMultipartUpload(t *testing.T, adapter block.Adapter, storageNamespace string) {
```

Comment:
Aborting multipart uploads is also important. 
It can be added in a different PR.

Reply:
Yes - not everything was covered in this PR, I'll open a separate issue

Reply:
https://github.com/treeverse/lakeFS/issues/5326